File: apps/cli/src/wire/loadrequest.ts
// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries
import { wire } from "../../../../libs/constants/src/index"

type LoadRequest = {
	wire: string
	type: wire.WireType
	collection: string
	fields: LoadRequestField[]
	conditions: LoadRequestCondition[]
	order: OrderDescription[]
	limit: number
	offset: number
}

type LoadRequestField = {
	id: string
}

type LoadRequestCondition =
	| LoadRequestLookupCondition
	| LoadRequestValueCondition

type LoadRequestLookupCondition = {
	field: string
	valuesource: "LOOKUP"
	lookupWire: string
	lookupField: string
}

type LoadRequestValueCondition = {
	field: string
	valuesource: "VALUE"
	value: string
}

type LoadRequestBatch = {
	wires: LoadRequest[]
}

type OrderDescription = {
	field: string
	desc: boolean
}

export { LoadRequest, LoadRequestField, LoadRequestBatch, LoadRequestCondition }

File: apps/platform/pkg/adapt/loadrequest.go
package adapt

// LoadRequest struct
type LoadRequest struct {
	Collection string                 `json:"collection"`
	Wire       string                 `json:"wire"`
	Type       string                 `json:"type"`
	Fields     []LoadRequestField     `json:"fields"`
	Conditions []LoadRequestCondition `json:"conditions"`
	Order      []LoadRequestOrder     `json:"order"`
	Limit      int                    `json:"limit"`
	Offset     int                    `json:"offset"`
}

// GetCollection function
func (lr *LoadRequest) GetCollection() string {
	return lr.Collection
}

// GetWire function
func (lr *LoadRequest) GetWire() string {
	return lr.Wire
}

// LoadRequestField struct
type LoadRequestField struct {
	ID     string             `json:"id"`
	Fields []LoadRequestField `json:"fields"`
}

// LoadRequestCondition struct
type LoadRequestCondition struct {
	Field       string      `json:"field"`
	Value       interface{} `json:"value"`
	ValueSource string      `json:"valueSource"`
	Type        string      `json:"type"`
	Operator    string      `json:"operator"`
	LookupWire  string      `json:"lookupWire"`
	LookupField string      `json:"lookupField"`
}

// LoadRequestOrder struct
type LoadRequestOrder struct {
	Field string `json:"field"`
	Desc  bool   `json:"desc"`
}
